man/apol.1
.TH apol 1
.SH NAME
apol \- SELinux policy analysis tool
.SH SYNOPSIS
.B apol
[OPTIONS] [POLICY ...]
.SH DESCRIPTION
.PP
.B apol
is a graphical tool that allows the user to inspect aspects of a SELinux policy.
The tool allows the user to browse policy components (types, classes, roles, users, etc.), rules (TE, RBAC, MLS), and file system contexts.
The tool also provides in depth analyses of domain transitions, information flows, and relabeling permissions.
.SH POLICY
.PP
.B
apol
supports loading a SELinux policy in one of four formats.
.IP "source"
A single text file containing policy source for versions 12 through 21. This file is usually named policy.conf.
.IP "binary"
A single file containing a monolithic kernel binary policy for versions 15 through 21. This file is usually named by version - for example, policy.20.
.IP "modular"
A list of policy packages each containing a loadable policy module. The first module listed must be a base module.
.IP "policy list"
A single text file containing all the information needed to load a policy, usually exported by SETools graphical utilities.
.PP
If a policy is not given on the command line then
.B
apol
will begin with none loaded.
.SH OPTIONS
.IP "-h, --help"
Print help information and exit.
.IP "-V, --version"
Print version information and exit.
.SH AUTHOR
This manual page was written by Jeremy A. Mowery <jmowery@tresys.com>.
.SH COPYRIGHT
Copyright(C) 2001-2007 Tresys Technology, LLC
.SH BUGS
Please report bugs via an email to setools-bugs@tresys.com.
.SH SEE ALSO
seinfo(1), sesearch(1), sechecker(1), indexcon(1)


NEWS
SETools 3.1:

SETools:

  * SETools now has an external dependency upon libsqlite3 >= 3.2.  The
    supplied configure script will enforce this dependency.

libsefs:

  * Rewrite of library to have proper namespaces and much more usable
    object-oriented design.

  * SWIG wrappers generated for this library if the appropriate
    configure flags are set.

findcon, searchcon:

  * Merge searchcon's functionality into findcon.  The searchcon tool
    has been removed from SETools.

indexcon, replcon:

  * Updated to use new libsefs design.

apol:

  * Updated to use new libsefs design.

==================================================

SETools 3.2:

libapol, libqpol, libesaudit, libpoldiff:

  * If --enable-swig-python is given during configure time, the build
    system will create Python SWIG wrappers for these libraries.

  * If --enable-swig-java is given during configure time, the build
    system will create Java SWIG wrappers for these libraries.

libpoldiff:

  * Provides ability to diff levels, categories, and range_transition
    statements.  Provides ability to diff modifications to a user's
    default level and permitted MLS range, assuming the policies are MLS.
    
apol, seaudit, sediffx:

  * Introduces 'policy list', a small text file that contains
    references to a base policy and any number of modules.  After
    selecting the base policy and modules in a tool's open policy
    dialog, click on 'Export' to write a policy list to disk.  That
    policy list then may be imported into the same tool or any other
    graphical SETools application.  This file also may be specified
    on the command line for all tools that load a policy.

sesearch, apol:

  * Provides full support for version 21 policy (i.e., object classes for
    range_transition statements).

sediff:

  * Shows MLS diffs.

sediffx:

  * Shows MLS diffs.

  * Provides option to show line numbers that contributed just to a
    specific AV rule's permission by clicking that AV rule's permissions.
    This is in addition to existing sediffx behavior that showed all lines
    associated with a particular rule difference.

  * Provides support for type joins and splits within the type remap dialog.

==================================================

SETools 3.1:

SETools:

  * All tools that open a policy now support loadable policy modules.
    Command line tools expect the first module to be a base module
    followed optionally by any other modules.  Graphical tools have
    a new open policy dialog to select a base module and any number of
    additional modules.

  * Release of RPM packages that are compatible with Fedora Core 5 and
    6.  The spec and support files are in packages/rpm.

libapol:

  * New class apol_policy_path_t to represent a base policy and any
    number of modules.  Use this whenever referring to the file or
    files constituting a policy.
  
libqpol:

  * Policy features such as attribute names or MLS can now be queried
    individally via qpol_policy_has_capability() rather than inferred
    by policy type and version.

  * New class qpol_module_t to represent a particular policy module
    prior to it being linked into a base policy (qpol_policy_t).

libseaudit:

  * Rewrite of library to have proper namespaces.  libseaudit is now
    fully documented and suitable for third-party users.

seaudit:

  * Rewrite to use new libseaudit.

  * Numerous tweaks to the interface to be more user friendly.

seaudit-report:

  * Rewrite to use new libseaudit.

sediffx:

  * Numerous tweaks to the interface to be more user friendly.

==================================================

SETools 3.0.1

SETools:

  * All code has been indented uniformly via the 'make indent' target.

SETools libraries:

  * All libraries now have a get_version() function.

libqpol:

  * Syntactic rule table is now off by default; it requires an
    explicit call to qpol_policy_build_syn_rule_table() to create it.

libapol:

  * apol_policy_t is now an opaque structure.  apol_permmap_t is no
    longer a public declaration.

  * avrule and terule queries now have full "syntactic" searching
    features.

apol & sesearch:

  * If loaded policy is source, the new syntactic search algorithm is used.

==================================================

SETools 3.0

SETools:

  * Moved entire project to autoconf build system.  This will detect
    dependencies correctly and make it easier to integrate with Linux
    distributions and their packaging systems.

SETools libraries:

  * Rewrite of back-end of SETools to use libsepol data structures.
    Most things should run notably faster.

  * All exported library functions have a standardized naming
    convention and are fully documented.  This will prove helpful for
    third-party developers integrating SETools into their own
    projects.

  * For policies version 15 or greater, domain transition analysis now
    takes into consideration setexec permission and/or type_transition
    rules.

apol:

  * Rework Tk interface to fit on 1024x768 displays.

  * Indirect matching of attributes now works with rule searches.

sediff:

  * New diffing algorithm yields significant speed improvements.

  * Can now diff neverallow and role_transition statements.

  * Streamlined results display - should be easier to read.

sechecker:

  * Updated module format and template for ease of extension.


=======================================================
May 1, 2006, Version 2.4

apol:
	File contexts tab now allows for MLS range searching if 
	the loaded database is from a MLS filesystem.
	Policy statistics dialog now shows MLS and ocontexts
	summaries.

libapol:
	Added support for loading base policies containing optionals.
	Added support for searching range transitions containing
	attributes.

libseaudit:
	Bugfix to support parsing FC5-style audit logs.

seaudit:
	Added date filters.

secmds:
	Added support to indexcon and searchcon for MLS filesytems.

apol/apol.c
/**
 *  @file
 *  Initialization code for apol.
 *
 *  @author Frank Mayer mayerf@tresys.com
 *  @author Jeremy A. Mowery jmowery@tresys.com
 *  @author Jason Tang jtang@tresys.com
 *
 *  Copyright (C) 2001-2007 Tresys Technology, LLC
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

#include <config.h>

#include <errno.h>
#include <getopt.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <tcl.h>
#include <tk.h>
#include "../libapol-tcl/apol_tcl_other.h"
#include <apol/policy-path.h>
#include <apol/util.h>

#define STARTUP_SCRIPT "apol.tcl"
#define COPYRIGHT_INFO "Copyright (C) 2001-2007 Tresys Technology, LLC"

static struct option const opts[] = {
	{"help", no_argument, NULL, 'h'},
	{"version", no_argument, NULL, 'V'},
	{NULL, 0, NULL, 0}
};

static apol_policy_path_t *path = NULL;

int Tcl_AppInit(Tcl_Interp * interp)
{
	char *script;

	if (Tcl_Init(interp) == TCL_ERROR) {
		return TCL_ERROR;
	}
	if (Tk_Init(interp) == TCL_ERROR) {
		return TCL_ERROR;
	}
	/* apolicy packagae initialization */
	if (apol_tcl_init(interp) == TCL_ERROR) {
		return TCL_ERROR;
	}

	/* find and start the TCL scripts */
	if (apol_tcl_get_startup_script(interp, STARTUP_SCRIPT) != TCL_OK) {
		fprintf(stderr, "Error finding TCL script: %s\n", Tcl_GetStringResult(interp));
		Tcl_DeleteInterp(interp);
		Tcl_Exit(1);
	}
	if (asprintf(&script, "%s/%s", interp->result, STARTUP_SCRIPT) < 0) {
		fprintf(stderr, "%s\n", strerror(errno));
		Tcl_DeleteInterp(interp);
		Tcl_Exit(1);
	}
	if (Tcl_EvalFile(interp, script) != TCL_OK) {
		fprintf(stderr, "Error while parsing %s: %s on line %d\n\nIf %s is set, make sure directory is correct.\n",
			script, Tcl_GetStringResult(interp), interp->errorLine, APOL_ENVIRON_VAR_NAME);
		Tcl_DeleteInterp(interp);
		Tcl_Exit(1);
	}
	free(script);

	/* if a policy file was provided on command line, open it */
	if (path != NULL) {
		char *policy_type = "monolithic";
		const char *primary_path = apol_policy_path_get_primary(path);
		Tcl_Obj *command[2], *path_objs[3], *o;
		path_objs[2] = Tcl_NewListObj(0, NULL);
		if (apol_policy_path_get_type(path) == APOL_POLICY_PATH_TYPE_MODULAR) {
			size_t i;
			const apol_vector_t *modules = apol_policy_path_get_modules(path);
			policy_type = "modular";
			for (i = 0; i < apol_vector_get_size(modules); i++) {
				const char *m = apol_vector_get_element(modules, i);
				o = Tcl_NewStringObj(m, -1);
				if (Tcl_ListObjAppendElement(interp, path_objs[2], o) == TCL_ERROR) {
					fprintf(stderr, "Error building initial load command: %s\n", Tcl_GetStringResult(interp));
					Tcl_DeleteInterp(interp);
					Tcl_Exit(1);
				}
			}
		}
		path_objs[0] = Tcl_NewStringObj(policy_type, -1);
		path_objs[1] = Tcl_NewStringObj(primary_path, -1);
		command[0] = Tcl_NewStringObj("::ApolTop::openPolicyFile", -1);
		command[1] = Tcl_NewListObj(3, path_objs);
		Tcl_EvalObjv(interp, 2, command, TCL_EVAL_GLOBAL);
		apol_policy_path_destroy(&path);
	}
	return TCL_OK;
}

static void usage(const char *program_name, bool_t brief)
{
	printf("Usage: %s [OPTIONS] [POLICY ...]\n\n", program_name);
	if (brief) {
		printf("\tTry %s --help for more help.\n\n", program_name);
		return;
	}
	printf("Policy Analysis tool for Security Enhanced Linux.\n\n");
	printf("   -h, --help              print this help text and exit\n");
	printf("   -V, --version           print version information and exit\n\n");
}

void print_version_info(void)
{
	printf("apol %s\n%s\n", VERSION, COPYRIGHT_INFO);
}

void parse_command_line(int argc, char **argv)
{
	int optc;
	bool_t help, ver;

	help = ver = FALSE;
	while ((optc = getopt_long(argc, argv, "hV", opts, NULL)) != -1) {
		switch (optc) {
		case 'h':
			help = TRUE;
			break;
		case 'V':
			ver = TRUE;
			break;
		default:
			usage(argv[0], TRUE);
			exit(1);
		}
	}
	if (help || ver) {
		if (help)
			usage(argv[0], FALSE);
		if (ver)
			print_version_info();
		exit(1);
	}
	if (argc - optind > 0) {
		apol_policy_path_type_e path_type = APOL_POLICY_PATH_TYPE_MONOLITHIC;
		char *policy_file = argv[optind];
		apol_vector_t *mod_paths = NULL;
		if (argc - optind > 1) {
			path_type = APOL_POLICY_PATH_TYPE_MODULAR;
			if (!(mod_paths = apol_vector_create(NULL))) {
				ERR(NULL, "%s", strerror(errno));
				exit(1);
			}
			for (optind++; argc - optind; optind++) {
				if (apol_vector_append(mod_paths, argv[optind])) {
					ERR(NULL, "Error loading module %s.", argv[optind]);
					apol_vector_destroy(&mod_paths);
					exit(1);
				}
			}
		} else if (apol_file_is_policy_path_list(policy_file) > 0) {
			path = apol_policy_path_create_from_file(policy_file);
		}
		if (!path)
			path = apol_policy_path_create(path_type, policy_file, mod_paths);
		if (!path) {
			ERR(NULL, "Error loading module %s.", argv[optind]);
			apol_vector_destroy(&mod_paths);
			exit(1);
		}
		apol_vector_destroy(&mod_paths);
	}
}

int main(int argc, char *argv[])
{
	Tcl_Interp *interp;

	parse_command_line(argc, argv);
	interp = Tcl_CreateInterp();
	/* Compute the full path name of the executable file from which
	 * the application was invoked and save it for Tcl's internal
	 * use. */
	Tcl_FindExecutable(argv[0]);

	/* Normally, the function 'Tk_MainEx(1, argv, Tcl_AppInit, interp);'
	 * would have been called instead of the following 'if' statement,
	 * however, this would cause apol to run wish in interactive mode,
	 * which is usually used for debugging purposes and this causes a
	 * strange interaction with normal shell operations (backgrounding,
	 * etc).
	 */
	if (Tcl_AppInit(interp) == TCL_OK)
		Tk_MainLoop();

	/* Exit after the event loop returns. */
	exit(0);
}
